src/extension_instance.ml

```diff
@@ -108,21 +108,68 @@ let stop_server t =
     else Promise.return ()
 ;;
 
-let check_ocaml_lsp_available sandbox =
-  let ocaml_lsp_version sandbox =
-    Sandbox.get_command sandbox "ocamllsp" [ "--version" ]
-  in
-  let cwd =
-    match Workspace.workspaceFolders () with
-    | [ cwd ] -> Some (cwd |> WorkspaceFolder.uri |> Uri.fsPath |> Path.of_string)
-    | _ -> None
+let suggest_to_run_dune_pkg_lock () =
+  let open Promise.Syntax in
+  let (_ : unit Promise.t) =
+    let+ maybe_choice =
+      Window.showWarningMessage
+        ~message:
+          "Dune Package Manager is selected as the active sandbox, but no lock file is \
+           present. Do you want to run dune pkg lock?"
+        ~choices:
+          [ ( "Generate lockfile"
+            , fun () ->
+                let (_ : Ojs.t option Promise.t) =
+                  Vscode.Commands.executeCommand
+                    ~command:Extension_consts.Commands.run_dune_pkg_lock
+                    ~args:[]
+                in
+                () )
+          ; ( "Pick another sandbox"
+            , fun () ->
+                let (_ : Ojs.t option Promise.t) =
+                  Vscode.Commands.executeCommand
+                    ~command:Extension_consts.Commands.select_sandbox
+                    ~args:[]
+                in
+                () )
+          ]
+        ()
+    in
+    Option.iter maybe_choice ~f:(fun f -> f ())
   in
-  Cmd.output ?cwd (ocaml_lsp_version sandbox)
-  |> Promise.Result.fold
-       ~ok:(fun (_ : string) -> ())
-       ~error:(fun (_ : string) ->
-         "Sandbox initialization failed: `ocaml-lsp-server` is not installed in the \
-          current sandbox.")
+  ()
+;;
+
+let check_ocaml_lsp_available (sandbox : Sandbox.t) =
+  match sandbox with
+  | Dune dune ->
+    let open Promise.Syntax in
+    let+ dune_lsp_present = Dune.detect_dune_ocamllsp dune in
+    if dune_lsp_present
+    then Ok ()
+    else (
+      let (_ : Ojs.t option Promise.t) =
+        Vscode.Commands.executeCommand
+          ~command:Extension_consts.Commands.install_dune_lsp
+          ~args:[]
+      in
+      Ok ())
```

Comment:
This function is meant to check if the lsp server is installed. If the chosen sandbox is not dune pkg management, it returns an error if the server not installed. This case is then handled by the caller, which in that case suggests to install the server. That seems like a good control flow! However, if the sandbox is dune pkg, this function directly installs the server and then returns Ok.

I'd suggest to have the same behavior for all sandboxes: If the server is not installed, this function returns an Err and the rest is handled by the caller. Does that sound good to you?